packages/yew-router/src/utils.rs

```diff
@@ -0,0 +1,62 @@
+use std::collections::HashMap;
+use wasm_bindgen::{JsCast, JsValue};
+
+pub fn base_url() -> Option<String> {
+    match yew::utils::document().query_selector("base") {
+        Ok(Some(base)) => {
+            let base = base
+                .unchecked_into::<web_sys::Element>()
+                .attributes()
+                .get_named_item("href")
+                .expect("base without href")
+                .value();
+            if base == "/" {
+                None
+            } else {
+                let base = base.strip_suffix("/").unwrap_or(&base);
+                Some(base.to_string())
+            }
+        }
+        _ => None,
+    }
+}
+
+pub fn build_path_with_base(to: &str) -> String {
+    let to = format!("{}{}", base_url().as_deref().unwrap_or(""), to);
+
+    let path = if to == "/" {
+        to
+    } else {
+        to.strip_suffix("/").map(|it| it.to_string()).unwrap_or(to)
+    };
+
+    path
+}
+
+pub fn get_query_params() -> HashMap<String, String> {
+    let url = web_sys::Url::new(&yew::utils::document().url().unwrap()).unwrap();
+
+    let iter = js_sys::try_iter(&JsValue::from(&url.search_params()))
+        .expect("try_iter failed")
+        .expect("try_iter failed")
+        .into_iter()
+        .map(|it| it.unwrap().unchecked_into::<js_sys::Array>().to_vec())
```

Comment:
Might be easier to just use
```rs
js_sys::Array::from(url.search_params().as_ref()).to_vec()
```

to get the `[key1, value1, key2, value2]` vector and then use `.chunks_exact(2)` to extract key/value pairs.

Reply:
That won't work. It's [this thing](https://discord.com/channels/701068342760570933/703449306497024049/822439572033634385).